### PR TITLE
Lowering of EXPONENT, FRACTION, SET_EXPONENT intrinsic

### PR DESCRIPTION
**Pull request by bolgrebolgre**
Implementation and tests of lowering for the exponent, fraction, and set_exponent intrinsic.

### CODE REVIEW

File: flang/lib/Lower/IntrinsicCall.cpp [hunk 1]
```diff
@@ -469,7 +469,10 @@ struct IntrinsicLibrary {
   mlir::Value genDprod(mlir::Type, llvm::ArrayRef<mlir::Value>);
   template <Extremum, ExtremumBehavior>
   mlir::Value genExtremum(mlir::Type, llvm::ArrayRef<mlir::Value>);
+  mlir::Value genExponent(mlir::Type, llvm::ArrayRef<mlir::Value>);
```
Comment (jeanPerier) [inline flang/lib/Lower/IntrinsicCall.cpp:472]: nit: `genExponent` should come before `genExtremum` to respect the alphabetical order (though it has no functional consequences here).

File: flang/runtime/numeric.cpp [hunk 2]
```diff
@@ -109,7 +109,7 @@ template <typename T> inline T SetExponent(T x, std::int64_t p) {
   } else if (x == 0) {
     return 0; // 0 -> 0
   } else {
-    int expo{std::ilogb(x)};
+    int expo{std::ilogb(x) + 1};
```
Comment (jeanPerier) [inline flang/runtime/numeric.cpp:112]: Can you also push this fix to `flang/runtime/numeric.cpp` directly into LLVM via phabricator ? We try as much as possible to keep LLVM the single point of truth for `flang/runtime`.

File: flang/lib/Lower/IntrinsicCall.cpp (hunk 1 above)
Comment (bolgrebolgre) [reply flang/lib/Lower/IntrinsicCall.cpp:472]: Yes I will change this. Thank you!

File: flang/runtime/numeric.cpp (hunk 2 above)
Comment (bolgrebolgre) [reply flang/runtime/numeric.cpp:112]: Yes I will do that. Thank you again!

**Review by mleair — APPROVED** [review]
LGTM

**Comment by schweitzpgi** [comment]
I suspect this new failure is related to this change.
```
  flang-Unit :: RuntimeGTest/./FlangRuntimeTests/Numeric.SetExponent
```

**Comment by schweitzpgi** [comment]
> > I suspect this new failure is related to this change.
> > ```
> >   flang-Unit :: RuntimeGTest/./FlangRuntimeTests/Numeric.SetExponent
> > ```
> 
> I will quickly look into this. When I tested with 'make check-flang or make check-flang-lower', did not see any failure. If I miss any tests, please let me know as well!

I ran `make check-flang`.

**Comment by bolgrebolgre** [comment]
> > > I suspect this new failure is related to this change.
> > > ```
> > >   flang-Unit :: RuntimeGTest/./FlangRuntimeTests/Numeric.SetExponent
> > > ```
> > 
> > 
> > I will quickly look into this. When I tested with 'make check-flang or make check-flang-lower', did not see any failure. If I miss any tests, please let me know as well!
> 
> I ran `make check-flang`.

So this is happening because of the change in flang/runtime/numeric.cpp:121 to have e number correctly.
Without this change, the output was always twice bigger than the output using other Fortran compilers (e.g., gfortran).

For example, the case of "Real<8>{3.0}, 0", the output of set_exponent(3.0, 0) from gfortran is 0.75, but the expected output set in RuntimeGTest/Numeric.cpp is 1.5. I probably misunderstood this, but I think the expected value from set_exponent should be adjusted with this change. 

What do you think? Any advice?

**Comment by schweitzpgi** [comment]
Looks like a bug in the runtime. I also get 0.75 out of gfortran.

I think an engineer here is working on a fix in llvm-project, btw.

**Pull request merged**